File: LeetCodeProblemsGo/LeetCode1653MinimumDeletionsToMakeStringBalanced/LeetCode1653MinimumDeletionsToMakeStringBalanced.go
package LeetCode1653MinimumDeletionsToMakeStringBalanced

import "fmt"

func minimumDeletions(s string) int {
	length := len(s)
	aAfter := make([]int, length)
	bBefore := make([]int, length)
	for i := 1; i < length; i++ {
		bBefore[i] = bBefore[i-1]
		if s[i-1] != 'b' {
			continue
		}
		bBefore[i]++
	}
	for i := length - 2; i >= 0; i-- {
		aAfter[i] = aAfter[i+1]
		if s[i+1] != 'a' {
			continue
		}
		aAfter[i]++
	}
	minRemove := length
	toRemove := 0

	for i := 0; i < length; i++ {
		toRemove = aAfter[i] + bBefore[i]
		minRemove = min(minRemove, toRemove)
	}

	return minRemove
}

func Calculate() {
	s := "ababaaaabbbbbaaababbbbbbaaabbaababbabbbbaabbbbaabbabbabaabbbababaa"
	fmt.Println(minimumDeletions(s))
}

File: LeetCodeProblemsCMake/Problems/LeetCode1653MinimumDeletionsToMakeStringBalanced.cpp
//
// Created by Saeed Jannati on 2/7/2026 A.
//

#include "LeetCode1653MinimumDeletionsToMakeStringBalanced.h"

using namespace std;

int LeetCode1653MinimumDeletionsToMakeStringBalanced::minimumDeletions(string s) {
    int length = s.length();
    int aAfter{};
    int bBefore{};
    for (auto c: s) {
        if (c == 'a') {
            aAfter++;
        }
    }
    int toRemove{};
    int minToRemove{length};
    for (auto c: s) {
        if (c == 'a') {
            aAfter--;
        }
        toRemove = aAfter + bBefore;
        minToRemove = min(minToRemove, toRemove);
        if (c == 'b') {
            bBefore++;
        }
    }
    return minToRemove;
}
